// bun-promptx/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  createPrompt: () => createPrompt,
  createSelection: () => createSelection
});
module.exports = __toCommonJS(src_exports);

// src/selection.ts
var import_bun_ffi3 = require("bun:ffi");

// src/ffi.ts
var import_bun_ffi = require("bun:ffi");
var import_meta = {};
var { platform, arch } = process;
var filename;
if (arch === "x64") {
  filename = `../release/promptx-${platform}-amd64.${import_bun_ffi.suffix}`;
} else {
  filename = `../release/promptx-${platform}-${arch}.${import_bun_ffi.suffix}`;
}
var location = new URL(filename, import_meta.url).pathname;
var { symbols } = (0, import_bun_ffi.dlopen)(location, {
  CreateSelection: {
    args: [import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.int],
    returns: import_bun_ffi.FFIType.ptr
  },
  CreatePrompt: {
    args: [import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.ptr, import_bun_ffi.FFIType.bool, import_bun_ffi.FFIType.int],
    returns: import_bun_ffi.FFIType.ptr
  },
  FreeString: {
    args: [import_bun_ffi.FFIType.ptr],
    returns: import_bun_ffi.FFIType.void
  }
});

// src/utils.ts
var import_bun_ffi2 = require("bun:ffi");
var utf8e = new TextEncoder();
function encode(data) {
  return utf8e.encode(data + "\0");
}
function toString(ptr3) {
  const str = new import_bun_ffi2.CString(ptr3);
  symbols.FreeString(str.ptr);
  return str.toString();
}

// src/selection.ts
function createSelection(items, options = {}) {
  const stringifiedItems = JSON.stringify(items.map((item) => {
    return {
      text: item.text,
      description: item.description || ""
    };
  }));
  const returnedPtr = symbols.CreateSelection(
    (0, import_bun_ffi3.ptr)(encode(stringifiedItems)),
    (0, import_bun_ffi3.ptr)(encode(options.headerText || "Select an item: ")),
    (0, import_bun_ffi3.ptr)(encode(options.footerText || "")),
    options.perPage || 5
  );
  const { selectedIndex, error } = JSON.parse(toString(returnedPtr));
  if (error !== "") {
    return {
      selectedIndex: null,
      error
    };
  }
  return {
    selectedIndex: Number(selectedIndex),
    error: null
  };
}

// src/prompt.ts
var import_bun_ffi4 = require("bun:ffi");
function createPrompt(prompt, options = {}) {
  const returnedPtr = symbols.CreatePrompt(
    (0, import_bun_ffi4.ptr)(encode(prompt)),
    (0, import_bun_ffi4.ptr)(encode(options.echoMode || "normal")),
    (0, import_bun_ffi4.ptr)(encode(options.validateOkPrefix || "")),
    (0, import_bun_ffi4.ptr)(encode(options.validateErrPrefix || "")),
    options.required ?? true,
    options.charLimit || 0
  );
  const { value, error } = JSON.parse(toString(returnedPtr));
  if (error !== "") {
    return {
      value: null,
      error
    };
  }
  return {
    value,
    error: null
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createPrompt,
  createSelection
});
